src/cpp/rtps/transport/TCPChannelResourceBasic.cpp

```diff
@@ -153,21 +155,16 @@ size_t TCPChannelResourceBasic::send(
 
     if (eConnecting < connection_status_)
     {
-        std::unique_lock<std::mutex> write_lock(write_mutex_);
-
-
+        std::vector<asio::const_buffer> buffers;
         if (header_size > 0)
         {
-            bytes_sent = socket_->send(asio::buffer(header, header_size), 0, ec);
-        }
-
-        if (!ec)
-        {
-            bytes_sent += socket_->send(asio::buffer(data, size), 0, ec);
+            buffers.push_back(asio::buffer(header, header_size));
         }
+        buffers.push_back(asio::buffer(data, size));
+        bytes_sent = asio::write(*socket_.get(), buffers, ec);
```

Comment:
In order to avoid the allocations of the temporary vector, I propose to change this block to:

```
        if (header_size > 0)
        {
            std::array<asio::const_buffer, 2> buffers;
            buffers[0] = asio::buffer(header, header_size);
            buffers[1] = asio::buffer(data, size);
            bytes_sent = asio::write(*socket_.get(), buffers, ec);
        }
        else
        {
            bytes_sent = asio::write(*socket_.get(), asio::buffer(data, size), ec);
        }
```